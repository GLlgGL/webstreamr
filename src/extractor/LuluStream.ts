import bytes from 'bytes';
import * as cheerio from 'cheerio';
import { NotFoundError } from '../error';
import { Context, Format, Meta, UrlResult } from '../types';
import {
  buildMediaFlowProxyExtractorStreamUrl,
  supportsMediaFlowProxy,
} from '../utils';
import { Extractor } from './Extractor';

/** @see https://github.com/Gujal00/ResolveURL/blob/master/script.module.resolveurl/lib/resolveurl/plugins/lulustream.py */
export class LuluStream extends Extractor {
  public readonly id = 'lulustream';

  public readonly label = 'LuluStream (via MediaFlow Proxy)';

  public override viaMediaFlowProxy = true;

  public supports(ctx: Context, url: URL): boolean {
    const supportedDomain = null !== url.host.match(/lulu/)
      || [
        '732eg54de642sa.sbs',
        'cdn1.site',
        'd00ds.site',
        'streamhihi.com',
      ].includes(url.host);

    return supportedDomain && supportsMediaFlowProxy(ctx);
  }

  public override normalize(url: URL): URL {
    const videoId = url.pathname.split('/').slice(-1)[0] as string;

    return new URL(`/e/${videoId}`, url);
  }

  protected async extractInternal(ctx: Context, url: URL, meta: Meta): Promise<UrlResult[]> {
    const headers = { Referer: meta.referer ?? url.href };

    const fileUrl = new URL(url.href.replace('/e/', '/d/'));
    const html = await this.fetcher.text(ctx, fileUrl, { headers });

    if (/No such file|File Not Found/.test(html)) {
      throw new NotFoundError();
    }

    const $ = cheerio.load(html);
    const title = $('h1').text().trim();

    const heightAndSizeMatch = html.match(/\d{3,}x(\d{3,}), ([\d.]+ ?[GM]B)/) as string[];

    const playlistUrl = await buildMediaFlowProxyExtractorStreamUrl(ctx, this.fetcher, 'LuluStream', url, headers);

    return [
      {
        url: playlistUrl,
        format: Format.hls,
        label: this.label,
        sourceId: `${this.id}_${meta.countryCodes?.join('_')}`,
        ttl: this.ttl,
        meta: {
          ...meta,
          title,
          bytes: bytes.parse(heightAndSizeMatch[2] as string) as number,
          height: parseInt(heightAndSizeMatch[1] as string),
        },
      },
    ];
  };
}
