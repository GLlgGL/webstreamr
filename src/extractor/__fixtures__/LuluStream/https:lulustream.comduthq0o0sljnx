<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="utf-8">
  <title>LuluStream</title>
  
  
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <!--<meta name="robots" content="noindex, nofollow, nosnippet, noarchive, noimageindex" />-->
  <meta name="theme-color" content="#FFF">
  <link rel="apple-touch-icon" href="/static/images/favicon/apple-touch-icon.png" sizes="180x180">
  <link rel="icon" href="/static/images/favicon/favicon-32x32.png" sizes="32x32">
  <link rel="icon" href="/static/images/favicon/favicon-16x16.png" sizes="16x16">
  <link rel="manifest" href="/static/images/favicon/manifest.json">
  <link rel="icon" href="/static/images/favicon/favicon.ico">
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin="crossorigin">
  <link href="https://fonts.googleapis.com/css2?family=Poppins:wght@400;500;700&display=swap" rel="stylesheet">
  <link href="/static/css/style.css?10" rel="stylesheet">

  <script src="/static/js/jquery-3.2.1.min.js"></script>
  <script src="/static/js/bootstrap.bundle.min.js"></script>
  <script src="/static/js/app.js"></script>
  <script src="/static/js/xupload.js?10"></script>
</head>
<body class="">
  
  <header class="header">
    <div class="container">
      <div class="row align-items-center flex-nowrap">
        <div class="col-auto flex-grow-1 flex-shrink-1">
          <div class="row align-items-center">
            <div class="col-auto d-lg-none">
              <div class="dropdown">
                <button class="btn icon-btn" data-bs-toggle="dropdown" data-bs-offset="0, 16">
                  <svg xmlns="http://www.w3.org/2000/svg" width="18" height="14">
                    <path fill-rule="evenodd" fill="currentcolor" d="M17 8H1a1 1 0 1 1 0-2h16a1 1 0 1 1 0 2Zm0-6H1a1 1 0 0 1 0-2h16a1 1 0 1 1 0 2ZM1 12h16a1 1 0 1 1 0 2H1a1 1 0 1 1 0-2Z" />
                  </svg>
                </button>
                <ul class="dropdown-menu dropdown-menu-end">
                  <a href="/?op=upload" class="dropdown-item">Upload</a>
                  <a href="/premium" class="dropdown-item">Premium</a>
                  <a href="/make_money" class="dropdown-item">Earn Money</a>
                  <a href="/api.html" class="dropdown-item">Service API</a>
                  <a href="/tos" class="dropdown-item">Terms of Service</a>
                  <a href="/faq" class="dropdown-item">About us</a>
                  <a href="/contact" class="dropdown-item">Contacts</a>
                </ul>
              </div>
            </div>
            <div class="col-auto">
              <a href="/" class="logo">
                <img src="/static/images/logo.svg" alt="">
              </a>
            </div>
            <div class="col-auto d-none d-lg-block">
              <nav class="nav">
                <a href="/make_money" class="nav-link">Earn Money</a>
                <a href="/premium" class="nav-link">Premium</a>
                <a href="/api.html" class="nav-link">Api Docs</a>
              </nav>
            </div>
          </div>
        </div>

        
        <div class="col-auto d-none d-lg-block">
          <a href="/login.html" class="btn login-btn me-1">Login</a>
          <a href="/?op=registration" class="btn btn-gradient reg-btn">Register</a>
        </div>
        <div class="col-auto d-lg-none">
          <div class="dropdown">
            <button class="btn icon-btn" data-bs-toggle="dropdown" data-bs-offset="0, 16">
              <svg xmlns="http://www.w3.org/2000/svg" width="20" height="20">
                <path fill-rule="evenodd" fill="currentcolor" d="M16.525 17.561c-.032.032-.062.063-.099.09A9.953 9.953 0 0 1 10 20a9.947 9.947 0 0 1-6.448-2.369c-.018-.014-.032-.029-.048-.043A9.976 9.976 0 0 1 0 10C0 4.486 4.486 0 10 0s10 4.486 10 10c0 3.02-1.351 5.726-3.475 7.561Zm-1.92-1.03A3.013 3.013 0 0 0 12.002 15H7.998a2.972 2.972 0 0 0-2.606 1.529 7.952 7.952 0 0 0 9.213.002ZM10 2c-4.411 0-8 3.589-8 8a7.96 7.96 0 0 0 1.889 5.151A4.962 4.962 0 0 1 7.998 13h4.005c1.657 0 3.184.84 4.103 2.157A7.962 7.962 0 0 0 18 10c0-4.411-3.589-8-8-8Zm0 10c-2.206 0-4-1.794-4-4s1.794-4 4-4 4 1.794 4 4-1.794 4-4 4Zm0-6c-1.103 0-2 .897-2 2s.897 2 2 2 2-.897 2-2-.897-2-2-2Z"
                />
              </svg>
            </button>
            <ul class="dropdown-menu dropdown-menu-end">
              <li>
                <a class="dropdown-item" href="/login.html">Login</a>
              </li>
              <li>
                <a class="dropdown-item" href="/?op=registration">Register</a>
              </li>
            </ul>
          </div>
        </div>
        


      </div>
    </div>
  </header>


  <main>

  <section class="section">
    <div class="container">
    
      <div class="alert alert-danger">
        No such file=uthq0o0sljnx
      </div>
    
    
  </div>
</section>
</main>
<footer class="footer">
  <div class="container">
    <div class="row">
      <div class="col-lg-4 d-flex flex-column justify-content-between">
        <div class="mb-3">

        </div>
        <p class="text-muted">Â© 2023 LuluStream. <br> All rights reserved.</p>
      </div>
      <div class="col-lg-3">
        <div class="nav flex-column mb-2">
          <a href="/tos" class="nav-link">Terms of service</a>
          <a href="/api.html" class="nav-link">API Documentation</a>
          <a href="/contact" class="nav-link">Contact Us</a>
        </div>
      </div>
      <div class="col-lg-3">
        <div class="nav flex-column mb-2">
          <a href="/premium" class="nav-link">Premium</a>
          <a href="/make_money" class="nav-link">Earn money</a>
          <a href="/check_files" class="nav-link">Link Checker</a>   
        </div>
      </div>
      <div class="col-lg-2 align-self-end text-end">
        <div class="mb-4">
          <a href="#" class="logo">
            <img src="/static/images/logo.svg" alt="">
          </a>
        </div>
      </div>
    </div>
  </div>
</footer>


<script>(function(){function c(){var b=a.contentDocument||a.contentWindow.document;if(b){var d=b.createElement('script');d.innerHTML="window.__CF$cv$params={r:'98a528aa0a1c9754',t:'MTc1OTc1MzcwOQ=='};var a=document.createElement('script');a.src='/cdn-cgi/challenge-platform/scripts/jsd/main.js';document.getElementsByTagName('head')[0].appendChild(a);";b.getElementsByTagName('head')[0].appendChild(d)}}if(document.body){var a=document.createElement('iframe');a.height=1;a.width=1;a.style.position='absolute';a.style.top=0;a.style.left=0;a.style.border='none';a.style.visibility='hidden';document.body.appendChild(a);if('loading'!==document.readyState)c();else if(window.addEventListener)document.addEventListener('DOMContentLoaded',c);else{var e=document.onreadystatechange||function(){};document.onreadystatechange=function(b){e(b);'loading'!==document.readyState&&(document.onreadystatechange=e,c())}}}})();</script>