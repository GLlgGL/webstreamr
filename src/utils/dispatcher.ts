import { socksDispatcher } from 'fetch-socks';
import { Agent, Dispatcher, interceptors, ProxyAgent } from 'undici';

const createBasicDispatcher = (): Dispatcher => {
  if (process.env['ALL_PROXY']) {
    const proxyUrl = new URL(process.env['ALL_PROXY']);
    if (proxyUrl.protocol === 'socks5:') {
      return socksDispatcher({ type: 5, host: proxyUrl.hostname, port: parseInt(proxyUrl.port) }, { allowH2: true });
    } else {
      return new ProxyAgent({ uri: proxyUrl.href, allowH2: true });
    }
  } else {
    return new Agent({ allowH2: true });
  }
};

export const createDispatcher = (): Dispatcher => {
  return createBasicDispatcher().compose(
    interceptors.dns(),
    interceptors.retry({ maxRetries: 3 }),
  );
};
