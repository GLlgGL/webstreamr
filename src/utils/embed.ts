import { unpack } from 'unpacker';

export const unpackEval = (html: string): string => {
  const evalMatch = html.match(/eval\(function\(p,a,c,k,e,d\).*\)\)/);
  if (!evalMatch) {
    throw new Error(`No p.a.c.k.e.d string found`);
  }

  return unpack(evalMatch[0]);
};

export const extractUrlFromPacked = (html: string, linkRegExps: RegExp[]): URL => {
  const unpacked = unpackEval(html);

  for (const linkRegexp of linkRegExps) {
    const linkMatch = unpacked.match(linkRegexp);
    if (linkMatch && linkMatch[1]) {
      return new URL(`https://${linkMatch[1].replace(/^(https:)?\/\//, '')}`);
    }
  }

  throw new Error(`Could not find a stream link in embed`);
};
