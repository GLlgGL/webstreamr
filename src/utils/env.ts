import * as os from 'node:os';
import { Request } from 'express';

export const envGet = (name: string): string | undefined => process.env[name];

export const envGetRequired = (name: string): string => {
  const value = envGet(name);
  if (!value) {
    throw new Error(`Environment variable "${name}" is not configured.`);
  }

  return value;
};

export const envGetAppId = (): string => process.env['MANIFEST_ID'] || 'webstreamr';

export const envGetAppName = (): string => process.env['MANIFEST_NAME'] || 'WebStreamr';

export const envIsProd = (): boolean => process.env['NODE_ENV'] === 'production';

export const isElfHostedInstance = (req: Request): boolean => req.host.endsWith('elfhosted.com');

export const getCacheDir = (): string => envGet('CACHE_DIR') ?? os.tmpdir();
