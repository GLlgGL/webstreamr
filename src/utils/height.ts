import { Context } from '../types';
import { CustomRequestInit, Fetcher } from './Fetcher';

export const guessHeightFromPlaylist = async (ctx: Context, fetcher: Fetcher, url: URL, init?: CustomRequestInit): Promise<number | undefined> => {
  const m3u8Data = await fetcher.text(ctx, url, init);

  const heights = Array.from(m3u8Data.matchAll(/\d+x(\d+)|(\d+)p/g))
    .map(heightMatch => heightMatch[1] ?? heightMatch[2])
    .filter(height => height !== undefined)
    .map(height => parseInt(height));

  return heights.length ? Math.max(...heights) : /* istanbul ignore next */ undefined;
};
