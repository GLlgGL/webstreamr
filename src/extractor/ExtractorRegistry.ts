import TTLCache from '@isaacs/ttlcache';
import winston from 'winston';
import { Context, CountryCode, UrlResult } from '../types';
import { isExtractorDisabled } from '../utils';
import { Extractor } from './Extractor';

export class ExtractorRegistry {
  private readonly logger: winston.Logger;
  private readonly extractors: Extractor[];
  private readonly urlResultCache: TTLCache<string, UrlResult[]>;

  public constructor(logger: winston.Logger, extractors: Extractor[]) {
    this.logger = logger;
    this.extractors = extractors;
    this.urlResultCache = new TTLCache({ max: 4096 });
  }

  public async handle(ctx: Context, url: URL, countryCode: CountryCode, title?: string | undefined): Promise<UrlResult[]> {
    const extractor = this.extractors.find(extractor => !isExtractorDisabled(ctx.config, extractor) && extractor.supports(ctx, url));
    if (!extractor) {
      return [];
    }

    const normalizedUrl = extractor.normalize(url);
    const cacheKey = extractor.viaMediaFlowProxy
      ? `${extractor.id}_${normalizedUrl}_${ctx.config.mediaFlowProxyUrl}`
      : `${extractor.id}_${normalizedUrl}`;

    let urlResults = this.urlResultCache.get(cacheKey) ?? [];
    if (this.urlResultCache.has(cacheKey)) {
      return urlResults.map(urlResult => ({ ...urlResult, ttl: this.urlResultCache.getRemainingTTL(cacheKey) }));
    }

    this.logger.info(`Extract stream URL using ${extractor.id} extractor from ${url}`, ctx);

    urlResults = await extractor.extract(ctx, normalizedUrl, countryCode, title);

    if (!urlResults.some(urlResult => urlResult.error) && extractor.ttl) {
      this.urlResultCache.set(cacheKey, urlResults, { ttl: extractor.ttl });
    }

    return urlResults;
  };
}
