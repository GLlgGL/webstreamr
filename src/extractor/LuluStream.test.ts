import winston from 'winston';
import { createTestContext } from '../test';
import { FetcherMock } from '../utils';
import { ExtractorRegistry } from './ExtractorRegistry';
import { LuluStream } from './LuluStream';

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
const extractorRegistry = new ExtractorRegistry(logger, [new LuluStream(new FetcherMock(`${__dirname}/__fixtures__/LuluStream`))]);

const ctx = createTestContext({ mediaFlowProxyUrl: 'https://mediaflow.test.org', mediaFlowProxyPassword: 'test' });

describe('LuluStream', () => {
  test('streamhihi d', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://streamhihi.com/d/mk9m58lz8ts6'))).toMatchSnapshot();
  });

  test('no such file', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://lulustream.com/e/uthq0o0sljnx'))).toMatchSnapshot();
  });
});
