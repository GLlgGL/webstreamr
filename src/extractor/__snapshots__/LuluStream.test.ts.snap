// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`LuluStream no such file 1`] = `[]`;

exports[`LuluStream streamhihi 1`] = `
[
  {
    "format": "hls",
    "label": "LuluStream (via MediaFlow Proxy)",
    "meta": {
      "bytes": 1288490188,
      "countryCodes": [],
      "height": 720,
      "title": "The Thursday Murder Club 2025 1080p NF WEB-DL DDP5 1 Atmos H 264-playWEB",
    },
    "sourceId": "lulustream_",
    "ttl": 900000,
    "url": "https://mediaflow.test.org/proxy/hls/manifest.m3u8?api_password=test&h_user-agent=Mozilla%2F5.0+%28Windows+NT+10.0%3B+Win64%3B+x64%29+AppleWebKit%2F537.36+%28KHTML%2C+like+Gecko%29+Chrome%2F136.0.0.0+Safari%2F537.36&h_referer=https%3A%2F%2Fstreamhihi.com%2Fe%2Fmk9m58lz8ts6&d=https%3A%2F%2Fh1gdnnp1dr11.tnmr.org%2Fhls2%2F02%2F02416%2Fqx1ysro2ny1o_h%2Fmaster.m3u8%3Ft%3DC1RS64a-9j6bSDO9TmHU9jIZpVb59jsWah7j-KgJkRg%26s%3D1759753708%26e%3D28800%26f%3D12081096%26i%3D0.3%26sp%3D0",
  },
]
`;

exports[`LuluStream streamhihi d 1`] = `
[
  {
    "format": "hls",
    "label": "LuluStream (via MediaFlow Proxy)",
    "meta": {
      "bytes": 1288490188,
      "countryCodes": [],
      "height": 720,
      "title": "The Thursday Murder Club 2025 1080p NF WEB-DL DDP5 1 Atmos H 264-playWEB",
    },
    "sourceId": "lulustream_",
    "ttl": 900000,
    "url": "https://mediaflow.test.org/proxy/hls/manifest.m3u8?api_password=test&h_user-agent=Mozilla%2F5.0+%28Windows+NT+10.0%3B+Win64%3B+x64%29+AppleWebKit%2F537.36+%28KHTML%2C+like+Gecko%29+Chrome%2F136.0.0.0+Safari%2F537.36&h_referer=https%3A%2F%2Fstreamhihi.com%2Fe%2Fmk9m58lz8ts6&d=https%3A%2F%2Fh1gdnnp1dr11.tnmr.org%2Fhls2%2F02%2F02416%2Fqx1ysro2ny1o_h%2Fmaster.m3u8%3Ft%3DC1RS64a-9j6bSDO9TmHU9jIZpVb59jsWah7j-KgJkRg%26s%3D1759753708%26e%3D28800%26f%3D12081096%26i%3D0.3%26sp%3D0",
  },
]
`;
